### PR TITLE
feat: Capture request, response headers and body

### PR DESCRIPTION
**Pull request by ernestii**
- [ ] test on android
- [ ] test on ios

### CODE REVIEW

File: src/instrumentations/xhr.ts [hunk 1]
```diff
@@ -727,7 +729,24 @@ export function instrumentXHR(config: XhrConfig) {
...
 
+  if (config.networkHeadersCapture) {
+    XMLHttpRequest.prototype.setRequestHeader = function (
+      this: XMLHttpRequest,
+      ...args
+    ) {
+      const [key, value] = args;
+      const xhrMem = _xhrMem.get(this);
+      if (xhrMem && key && value) {
+        const newKey = `http.request.header.${key.toLowerCase()}`;
```
Comment (wrn14897) [inline src/instrumentations/xhr.ts]: Let's try to make sure the headers extraction matches what otel does in their http instrumentation (https://github.com/open-telemetry/opentelemetry-js/blob/82b7526b028a34a23936016768f37df05effcd59/experimental/packages/opentelemetry-instrumentation-http/src/utils.ts#L583)

File: src/instrumentations/xhr.ts [hunk 2]
```diff
@@ -742,6 +761,14 @@ export function instrumentXHR(config: XhrConfig) {
           taskCounter.increment();
           xhrMem.sendStartTime = hrTime();
           currentSpan.addEvent(EventNames.METHOD_SEND);
+          if (config.networkBodyCapture) {
+            currentSpan.setAttribute(
+              'http.request.body',
+              typeof requestBody === 'string'
+                ? requestBody
+                : JSON.stringify(requestBody)
```
Comment (wrn14897) [inline src/instrumentations/xhr.ts]: better to handle the case when stringify throws. also I wonder if we want to set a cap on body size ?

File: src/instrumentations/xhr.ts [hunk 3]
```diff
@@ -753,7 +780,28 @@ export function instrumentXHR(config: XhrConfig) {
                 }
               }
             } else if (this.readyState === XMLHttpRequest.DONE) {
-              endSpan(EventNames.EVENT_READY_STATE_CHANGE, this);
+              if (config.networkBodyCapture && this.responseType === 'blob') {
+                try {
```
Comment (wrn14897) [inline src/instrumentations/xhr.ts]: curious what this try-catch is for

Comment (ernestii) [reply src/instrumentations/xhr.ts]: probably leftover from experimenting, removed

File: src/instrumentations/xhr.ts (hunk 2 above)
Comment (ernestii) [reply src/instrumentations/xhr.ts]: added!

File: src/instrumentations/xhr.ts (hunk 1 above)
Comment (ernestii) [reply src/instrumentations/xhr.ts]: updated!

File: src/instrumentations/xhr.ts [hunk 4]
```diff
@@ -753,7 +807,24 @@ export function instrumentXHR(config: XhrConfig) {
...
             } else if (this.readyState === XMLHttpRequest.DONE) {
-              endSpan(EventNames.EVENT_READY_STATE_CHANGE, this);
+              if (config.networkBodyCapture && this.responseType === 'blob') {
+                new Response(this.response)
+                  .text()
+                  .then((text) => {
+                    currentSpan.setAttribute('http.response.body', text);
+                  })
+                  .finally(() => {
+                    endSpan(EventNames.EVENT_READY_STATE_CHANGE, this);
```
Comment (wrn14897) [inline src/instrumentations/xhr.ts:817]: just double check. the span would never end before this ?

**Review by wrn14897 — APPROVED** [review]
Super excited about this! 🚢 🚢 🚢

File: src/instrumentations/xhr.ts [hunk 5]
```diff
@@ -47,11 +47,14 @@ import type { PropagateTraceHeaderCorsUrls } from '@opentelemetry/sdk-trace-web/
 
 const parseUrl = (url: string) => new URL(url);
 
+const MAX_BODY_LENGTH = 5 * 1024; // 5KB
```
Comment (wrn14897) [inline src/instrumentations/xhr.ts]: 5k seems a bit too strict. maybe something like 2MB to start with is okay

File: src/instrumentations/xhr.ts (hunk 4 above)
Comment (ernestii) [reply src/instrumentations/xhr.ts:817]: it shouldn't

File: src/instrumentations/xhr.ts (hunk 5 above)
Comment (ernestii) [reply src/instrumentations/xhr.ts]: updated

**Pull request merged**